Standardize function signature of compare_images.

mkcor opened the pull request:
## Description

This PR is a follow-up to #7255.

## Checklist

<!-- Before pull requests can be merged, they should provide: -->

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

Summarize the introduced changes in the code block below in one or a few sentences. The
summary will be included in the next release notes automatically:

```release-note
In `skimage.util.compare_images`, deprecate the parameter `image2`.
Instead use `image0`, `image1` to pass the compared images. Furthermore,
all other parameters will be turned into keyword-only parameters once
the deprecation is complete.
```

mkcor committed:
Remove obsolete instruction for documenting changes

mkcor committed:
Merge branch 'main' of github.com:scikit-image/scikit-image

mkcor committed:
Deprecate parameter image2 and limit to two positional arguments

mkcor committed:
Add test for replaced parameters

mkcor committed:
Use pytest

mkcor committed:
Cover all cases

mkcor committed:
Remove unnecessary default value

mkcor committed:
Add directives for future release

mkcor committed:
Ensure valid existing code does not break

mkcor committed:
Handle deprecation in a custom way

mkcor committed:
Capture all deprecation warnings

mkcor committed:
Fix typo in test name

mkcor committed:
Update instructions for future release

lagru commented on skimage/util/compare.py:
```diff
@@ -1,27 +1,52 @@
 
-def compare_images(image1, image2, method='diff', *, n_tiles=(8, 8)):
+def compare_images(
+    image0=None, image1=None, image2=DEPRECATED, *, method='diff', n_tiles=(8, 8)
```
While using `image0=None` has the advantage of being simple, I'd prefer a wrapper around `compare_images` here. That way, we can still perform the necessary checks, but `functools.wraps(func)` should ensure that IDE's and code completion pick up the actual signature I'm suggesting below
```suggestion
    image0, image1, image2=DEPRECATED, method='diff', *, n_tiles=(8, 8)
```
Note you can't yet move the `*` forward because we still want the old API to work during the deprecation cycle.

mkcor replied:
Ok. Should I write a custom wrapper then? Just to be sure, do you agree that a) I can't use `deprecate_parameter` here and b) existing code like `compare_images(image1=a, image2=b)` should still work during the deprecation cycle as well?

lagru replied:
a) yes and b) yes. I think a specific wrapper should be able to address all concerns in this case. I think something like
```python

def _rename_image_params(func):

    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        # check / fix kwargs and warn if `image2` is used
        return func(*args, **kwargs)

    return wrapper
``` 
should work.

mkcor replied:
Thank you, @lagru. I'll update the function and tests accordingly.

lagru commented on skimage/util/tests/test_compare.py:
```diff
@@ -20,8 +19,24 @@ def test_compare_images_diff():
+    img1[3:8, 3:8] = 255
+    img2 = np.zeros_like(img1)
+    img2[3:8, 0:8] = 255
+    with pytest.warns(FutureWarning):
```
Might at least use `match=".*image[01]` here and elsewhere.

mkcor replied:
I'm not sure... I removed the case where input images are passed as kwargs following the new API (e0d5929d4e321c1ae), so we're not quite cycling over all image0-image1 combinations when expecting a warning. Or maybe my command of regular expressions is not up to the mark :wink:

lagru reviewed (commented):
Thanks for tackling this @mkcor. Especially the somewhat complex documentation of the deprecation. :sweat_smile:

lagru commented on skimage/util/compare.py:
```diff
@@ -1,27 +1,52 @@
         Details are provided in the note section.
+
+    .. versionchanged:: 0.23
+            This parameter is now keyword-only.
```
What would `change_parameter` do? Maybe for a start we can focus on this case for now after which we have a better understanding and are already part-way there if we want to create `change_parameter`; at least that's how https://github.com/scikit-image/scikit-image/pull/7256 started.

mkcor committed:
Merge branch 'main' of github.com:scikit-image/scikit-image

mkcor committed:
Keep signature compatible with old API

Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

mkcor committed:
Merge branch 'main' of github.com:scikit-image/scikit-image

mkcor committed:
Update TODO

mkcor committed:
Merge branch 'main' into image0-image1

mkcor committed:
Merge branch 'main' into image0-image1

mkcor committed:
Remove test for new API

mkcor committed:
Remove warning test cases for new API

mkcor committed:
Write custom deprecation class

mkcor commented:
@lagru I've drawn from your class `deprecate_parameter` but, as you will see, it's a bit ugly, hard-coded, etc. :see_no_evil: I rushed to submit this, in case it could make it into v0.23. Cc @stefanv

lagru commented on skimage/util/compare.py:
```diff
@@ -1,27 +1,143 @@
 
-def compare_images(image1, image2, method='diff', *, n_tiles=(8, 8)):
+@_rename_image_params("image2", start_version="0.23", stop_version="0.25")
+def compare_images(image0, image1, image2=DEPRECATED, method='diff', *, n_tiles=(8, 8)):
```
I'm sorry but unfortunately I think looking at `deprecate_parameter` has lead you on a wrong path and over-complicated things. And also breaks

```python
a = np.ones(10)
b = np.zeros(10)
ski.util.compare_images(a, b, "blend")
# warns but still uses method="diff"
```

which used to work before and should continue working during the deprecation. I think we can't use an argument with `=DEPRECATED` here.

Except for that I think you got quite close, especially with the idea to first turn all `args` into `kwargs` to more easily handle them. What do you think about the draft and decorator below? It should work for all cases unless I've forgotten a a way to abuse the signature. It also turns `method` into a keyword-only parameter.

```python
def _rename_image_params(func):

    @functools.wraps(func)
    def wrapper(*args, **kwargs):

        # Turn all into keyword parameters
        for i, (value, param) in enumerate(
            zip(args, ["image0", "image1", "method", "n_tiles"])
        ):
            if i >= 2:
                warnings.warn("method will become keyword-only")
            if param in kwargs:
                raise ValueError(
                    f"{param} given as positional and keyword argument"
                )
            else:
                kwargs[param] = value
        args = tuple()

        # Account for `image2` if given
        if "image2" in kwargs:
            warnings.warn("used deprecated image2")

            # Safely move `image2` to `image1` if that's empty
            if "image1" in kwargs:
                # Safely move `image1` to `image0`
                if "image0" in kwargs:
                    raise ValueError(
                        "three images given, `image0`, `image1` and `image2`"
                    )
                kwargs["image0"] = kwargs.pop("image1")
            kwargs["image1"] = kwargs.pop("image2")

        return func(*args, **kwargs)

    return wrapper
```

Of course it doesn't have the nice convenience stuff of modifying the docstring and warnings and error messages are drafts and need to be polished. But I think this deprecation is so special that we should just edit the docstring by hand.

mkcor replied:
> I'm sorry but unfortunately I think looking at `deprecate_parameter` has lead you on a wrong path and over-complicated things.

No worries! I did this quickly and thought I would iterate afterwards.

> And also breaks
> 
> ```python
> a = np.ones(10)
> b = np.zeros(10)
> ski.util.compare_images(a, b, "blend")
> # warns but still uses method="diff"
> ```
> 
> which used to work before and should continue working during the deprecation.

Pardon me? I just tried locally and it works; it warns, indeed, and it returns
```py
array([0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5])
```
just like current main. `method="diff"` returns
 ```py
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])
```
instead. I just noticed that `method="checkerboard"` errors because `a` and `b` are 1-dimensional, which doesn't work with
https://github.com/scikit-image/scikit-image/blob/e453a53d23a980408365d35579008a588868ffe0/skimage/util/compare.py#L48

So we should probably improve the function by checking, in the `method="checkerboard"` case, that ndim == 2 :grimacing:

mkcor replied:
> So we should probably improve the function by checking, in the `method="checkerboard"` case, that ndim == 2 😬

Done db1b7740e21637a3. If you prefer, I can cherry-pick this commit and submit it as a separate PR.

lagru replied:
Ah, you are right, it does handle this. I chose my example poorly and got confused by the `image2=DEPRECATED` as the third positional arg. 

... which brings me to the point that we should probably remove it and make it so that the function shows the expected new signature as un-ambiguously as possible.

lagru commented on skimage/util/compare.py:
```diff
@@ -1,27 +1,143 @@
+                kwargs['image0'] = args[0]
+                if len(args) > 1:
+                    kwargs['image1'] = args[1]
+                    if len(args) > 2 and args[len(args) - 1] in [
+                        'diff',
+                        'blend',
+                        'checkerboard',
+                    ]:
```
The wrapper shouldn't recognize the `arg` for `method` by value but by position. The function will/should handle if a wrong value is passed as `method`.

mkcor replied:
Yes, the function does handle it:
https://github.com/scikit-image/scikit-image/blob/e453a53d23a980408365d35579008a588868ffe0/skimage/util/compare.py#L59-L62
Basically I wanted to cover both cases, where a user might call:
```py
compare_images(a, b, 'diff') 
```
and another one might call:
```py
compare_images(a, b, c, 'diff') 
```
which should definitely error -- but this should also be taken care of by the function, right? I realize that my RuntimeError("Use `image0, image1` to pass the two input images.") isn't catching it anyway, it's only catching
```py
compare_images(image0=a, image1=b, image2=c, ...) 
```

lagru replied:
That's why I think we shouldn't use the value of the 3rd and 4th argument to determine which one is `method`... just don't use `image2=DEPRECATED` and we can simply use the position reliably? Right now it seems even harder to think of all edge cases. E.g. consider
```python
import skimage as ski
a = np.ones((10, 10)) * 3
b = np.zeros((10, 10))
ski.util.compare_images(a, b, "unknown")
```
This will neither warn nor error but will simply return as if the user requested `compare_images(a, b)` (default `method=diff`). It should definitely error.

lagru replied:
Replying to your question: `compare_images(a, b, c, 'diff') ` should definitely error but it's not an expected use case of the old or new API so users shouldn't really encounter this case unless they try to mess around.

mkcor replied:
> Replying to your question: `compare_images(a, b, c, 'diff') ` should definitely error but it's not an expected use case of the old or new API so users shouldn't really encounter this case unless they try to mess around.

Right; there's definitely no reason to 'handle' it within the function itself and perhaps not even through the decorator...

mkcor replied:
> That's why I think we shouldn't use the value of the 3rd and 4th argument to determine which one is `method`... just don't use `image2=DEPRECATED` and we can simply use the position reliably? Right now it seems even harder to think of all edge cases. E.g. consider
> 
> ```python
> import skimage as ski
> a = np.ones((10, 10)) * 3
> b = np.zeros((10, 10))
> ski.util.compare_images(a, b, "unknown")
> ```
> 
> This will neither warn nor error but will simply return as if the user requested `compare_images(a, b)` (default `method=diff`). It should definitely error.

Ouch, yes. Maybe we should also check the args' types then? Let me try something based on your [suggestion](https://github.com/scikit-image/scikit-image/pull/7322/files#r1533795486) and ensure that yet this other edge case is covered.

mkcor replied:
In the end, I was lazy and went for 9b9f757ecc619be (not trying to figure out what the third positional argument was meant to be...).

mkcor committed:
Ensure images are 2D for method=checkerboard

mkcor committed:
Merge branch 'main' into image0-image1

lagru commented:
Sorry, @mkcor I don't want to come of as devaluing your approach. I'm happy with your solution as long as it gets all edge cases. :) There are a lot of edge cases to keep in mind and the longer I think about it the more I'm undecided which signature is actually least confusing to users during the deprecation.

Is it (a)
```python
def compare_images(image0, image1, image2=DEPRECATED, method='diff', *, n_tiles=(8, 8))
```
because it's easy to see that `image2` is deprecated but harder to see how `method` is affected? It's subtle to notice that not being able to use `image2` positionally also means `method` can no longer be used this way.

Or is it (b)
```python
def compare_images(image0, image1, *, method='diff', n_tiles=(8, 8))
```
because this one makes it easy to see that `method` is now a keyword-only parameter.

mkcor commented:
> Sorry, @mkcor I don't want to come off as devaluing your approach.

All good, @lagru! I think my code wasn't very elegant, so I definitely welcome your inputs. :heart: 

Now, maybe this kind of code doesn't have to be the most elegant anyway, since we will remove it two releases later?

I haven't looked closely at your alternative yet, just because potential additional edge cases came up... 

> I'm happy with your solution as long as it gets all edge cases. :) There are a lot of edge cases to keep in mind

Exactly, that's the point.

> and the longer I think about it the more I'm undecided which signature is actually least confusing to users during the deprecation.
> 
> Is it (a)
> 
> ```python
> def compare_images(image0, image1, image2=DEPRECATED, method='diff', *, n_tiles=(8, 8))
> ```
> 
> because it's easy to see that `image2` is deprecated but harder to see how `method` is affected? It's subtle to notice that not being able to use `image2` positionally also means `method` can no longer be used this way.

Exactly, that's the main problem with this deprecation (more so than `image1, image2` changing to `image0, image1`).

> Or is it (b)
> 
> ```python
> def compare_images(image0, image1, *, method='diff', n_tiles=(8, 8))
> ```
> 
> because this one makes it easy to see that `method` is now a keyword-only parameter.

This was my initial move, but you said (and rightfully so) that the old API (where `method` may be used positionally) should still work during the deprecation cycle... What about (c)

```python
def compare_images(image0, image1, method='diff', *, n_tiles=(8, 8))
```
?

mkcor committed:
Add edge case with positional args only

Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

mkcor committed:
Account for method-related warning

mkcor commented:
Not sure we're covering all edge cases, but this implementation is definitely lighter! Thanks, @lagru. :relaxed:

mkcor committed:
Refactor decorator

Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

mkcor committed:
Merge branch 'main' into image0-image1

lagru commented on skimage/util/compare.py:
```diff
@@ -1,27 +1,100 @@
+
+
+@_rename_image_params
+def compare_images(image0, image1, method='diff', *, n_tiles=(8, 8)):
```
```suggestion
def compare_images(image0, image1, *, method='diff', n_tiles=(8, 8)):
```
Because the wrapper turns all given `args` into `kwargs` we can already move the asterisk (marking keyword-only arguments) to the final position.

mkcor replied:
Oh, right! Thanks, @lagru.

lagru commented on skimage/util/compare.py:
```diff
@@ -1,27 +1,100 @@
+
+        .. versionchanged:: 0.23
+            `image1` changed from being the name of the first image to that of
+            the second image.
```
Thanks for documenting the change in such detail! Though, what do you think about simplifying it somewhat so it's easier to grok?

```suggestion
    image0, image1 : ndarray, shape (M, N)
        Images to process, must be of the same shape.

        .. versionchanged:: 0.23
            `image1` and `image2` where renamed to `image0` and `image1`
            respecitively.
```

lagru commented on skimage/util/tests/test_compare.py:
```diff
@@ -1,27 +1,47 @@
+def test_compare_images_ValueError_args():
+    a = np.ones((10, 10)) * 3
+    b = np.zeros((10, 10))
+    with pytest.warns(FutureWarning):
+        with pytest.raises(ValueError):
+            compare_images(a, b, "unknown")
```
```suggestion
      with pytest.raises(ValueError):
          compare_images(a, b, method="unknown")
```
I'd keep the tests focused and independent from each other; you're already testing the deprecation in another test. :)

mkcor replied:
Ok, you're right. I'll just edit the indentation of your edit.

lagru commented on skimage/util/tests/test_compare.py:
```diff
@@ -1,27 +1,47 @@
     expected_result[3:8, 0:3] = 1
-    result = compare_images(img1, img2, method='diff')
-    assert_array_equal(result, expected_result)
+    with pytest.warns(FutureWarning):
+        result = compare_images(img1, img2, 'diff')
```
```suggestion
    result = compare_images(img1, img2, method='diff')
```

lagru committed:
Correct and test stacklevel of deprecation warnings

The previous default stacklevel wouldn't have shown a confusing origin
of the warning to users.

lagru commented on TODO.txt:
```diff
@@ -8,6 +8,11 @@ Version 0.25
+* In `skimage/util/compare.py`, remove deprecated parameter `image2` as
+  well as `test_compare_images_replaced_param` in `skimage/util/tests/test_compare.py`
+  (and all `pytest.warns(FutureWarning)` context managers there).
+  Make all arguments from `method` (inclusive) keyword-only:
+  `compare_images(image0, image1, *, method='diff', ...)`.
```
```suggestion
```
Not needed, see other comment.

lagru reviewed (approved):
@mkcor thanks for addressing the discussion! I've pushed https://github.com/scikit-image/scikit-image/pull/7322/commits/0756a563c27cb8757d513025bea58192430f4f5c to correct and test the stacklevels. I'm happy with this but left a few minor concerns to make the completing the deprecation easier.

mkcor committed:
Update TODO.txt

Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

mkcor committed:
Apply suggestions from code review

Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

mkcor committed:
Fix typos

mkcor commented:
Thank you for the review and for 0756a563c27cb8757d513025bea58192430f4f5c, @lagru! I just included minor fixes in the docstring (770499a42a66c5058551738bc374601a52255842).

lagru committed:
Update deprecation messages to 0.24

lagru committed:
Merge branch 'main' into pr/7322_image0-image1

mkcor commented:
@lagru thank you for taking care of 17cfe6477db9ead20d76a4c25526619874fd31aa! :pray:

lagru committed:
Use correct indent for versionchanged directive

and reduce visual noise. The first versionchanged already informs users
that image2 was deprecated.

lagru commented on skimage/util/compare.py:
```diff
@@ -1,23 +1,85 @@
     """
     Return an image showing the differences between two images.
 
     .. versionadded:: 0.16
```
@mkcor I corrected a minor indent issue in https://github.com/scikit-image/scikit-image/pull/7322/commits/9a10712392ba8bd2656cc47215eb4de8a1513441. Also, I'm thinking about how we can reduce the visual noise these kind of directives create in our docs. E.g. this directive I'm commenting on doesn't seem all that useful. We should probably reserve using them for breaking changes when they are absolutely necessary to avoid user confusion. And do we plan on removing them ever? Curious what you think. :)

Let's merge once the CI is green.

lagru commented:
No idea about the new [error](https://github.com/scikit-image/scikit-image/actions/runs/8819647825/job/24213215782?pr=7322#step:5:208) yet:
```
../meson.build:1:0: ERROR: Compiler /usr/bin/clang cannot compile programs.
```
But it seems unrelated since this PR doesn't touch any part of the infrastructure..

jarrodmillman reviewed (approved):
Will merge once #7407 is sorted out and merged. Thanks!!

lagru committed:
Merge branch 'main' into pr/7322_image0-image1
